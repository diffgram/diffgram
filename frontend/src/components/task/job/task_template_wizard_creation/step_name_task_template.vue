<template>
  <v-container fluid data-cy="task-template-step-name">
    <h1 data-cy="wizard-title" class="font-weight-medium text--primary mb-8">
      <v-icon color="primary" class="mr-4" size="48">mdi-brush</v-icon>
      {{ title }}
    </h1>
    <v_error_multiple :error="error"></v_error_multiple>
    <p class="text--primary" data-cy="wizard-name-subtitle">
      {{ message }}
    </p>

    <v-text-field label="Name"
                  data-cy="task-template-name-input"
                  v-model="job.name">
    </v-text-field>

    <wizard_navigation
      @next="on_next_button_click"
      :loading_next="loading"
      :disabled_next="loading"
      :skip_visible="false"
      :back_visible="false"
    >

    </wizard_navigation>
  </v-container>

</template>

<script lang="ts">

import axios from '../../../../services/customInstance';

import Vue from "vue";

export default Vue.extend({
    name: 'step_name_task_template',
    props: [
      'project_string_id',
      'job',
      'title',
      'message'
    ],

    components: {},

    data() {
      return {
        error: {},
        loading: false,
      }
    },
    created() {

    },

    computed: {},
    methods: {
      verify_name: function () {
        if (!this.$props.job.name || this.$props.job.name === '') {
          this.error = {
            name: 'Name must not be empty.'
          }
          return false
        }
        return true
      },
      on_next_button_click: async function () {
        this.error = {};
        let name_ok = this.verify_name();
        if (name_ok) {
          await this.job_new();
          this.$emit('next_step');
        }
      },
      job_new: async function () {
        if (this.job.id != undefined) {
          return
        }
        this.loading = true
        this.job_new_error = {}
        let job = {...this.$props.job};
        // QUESTION
        // Not clear if we want to do this as a computed property or...

        // careful to update with share property that's seperate due to dict thing.
        const url = `/api/v1/project/${this.project_string_id}/job/new`

        try {
          const response = await axios.post(url, job)
          // Handle job hash / draft / job status
          if (response.data.log.success == true) {
            this.job.id = response.data.job.id;
            this.loading = false
            this.success_action = 'Job created successfully.'
            return response
          }

        } catch (error) {
          console.error(error);
          this.loading = false
          this.error = this.$route_api_errors(error)
          return error;
        }
      },

    }
  }
) </script>
