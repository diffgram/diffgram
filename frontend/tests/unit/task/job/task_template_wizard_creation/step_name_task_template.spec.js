import {shallowMount, createLocalVue} from "@vue/test-utils";
import Vuex from 'vuex'
import axios from '../../../../../src/services/customInstance'
import step_name_task_template
  from "../../../../../src/components/task/job/task_template_wizard_creation/step_name_task_template";
import VueRouter from 'vue-router'

const localVue = createLocalVue();
localVue.use(Vuex)
localVue.use(VueRouter)
const router = new VueRouter()

let job = {
  name: '',
  label_mode: 'closed_all_available',

  loading: false,
  passes_per_file: 1,
  share_object: {
    // TODO this may fail for org jobs? double check this.
    'text': String,
    'type': 'project'
  },
  share: 'project',
  instance_type: 'box', //"box" or "polygon" or... "text"...
  permission: 'all_secure_users',
  field: 'Other',
  category: 'visual',
  attached_directories_dict: {attached_directories_list: []},
  type: 'Normal',
  connector_data: {},
  // default to no review while improving review system
  review_by_human_freqeuncy: 'No review', //'every_3rd_pass'
  td_api_trainer_basic_training: false,
  file_handling: "use_existing",
  interface_connection: undefined,
  member_list_ids: ["all"]
}
describe("step_name_task_template.vue", () => {
  let actions;
  let store;

  beforeEach(() => {
    actions = {
      actionClick: jest.fn(),
      actionInput: jest.fn()
    }
    store = new Vuex.Store({
      actions
    })
  });


  it("Renders component correctly.", () => {
    const wrapper = shallowMount(step_name_task_template, {
      store,
      localVue,
      propsData: {
        job: job
      }
    });
    expect(wrapper).toMatchSnapshot();
  });

  it("Populates error data when calling verify_name().", () => {
    const wrapper = shallowMount(step_name_task_template, {
      store,
      localVue,
      propsData: {
        job: job
      }
    });
    wrapper.vm.verify_name();
    expect(wrapper.vm.error).toMatchObject({
      name: 'Name must not be empty.'
    })
  });

  it("Emits event when calling on_next_button_click()", async () => {
    const wrapper = shallowMount(step_name_task_template, {
      store,
      localVue,
      propsData: {
        job: {
          ...job,
          id: 5,
          name: 'test'
        }
      }
    });
    await wrapper.vm.on_next_button_click()
    expect(wrapper.emitted().next_step.length).toBe(1);
  });

  it("Makes request to /job/new when calling job_new()", () => {
    jest.mock('axios', () => ({
      post: jest.fn(() => {})
    }))
    const wrapper = shallowMount(step_name_task_template, {
      store,
      localVue,
      propsData: {
        job: job,
        project_string_id: 'test'
      }
    });
    let url = `/api/v1/project/test/job/new`;
    const spy = jest.spyOn(axios, 'post')
    wrapper.vm.job_new()
    expect(spy).toHaveBeenCalledWith(url, job);
  });



});
