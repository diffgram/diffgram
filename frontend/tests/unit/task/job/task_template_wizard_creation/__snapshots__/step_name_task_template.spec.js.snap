// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`step_name_task_template.vue Renders component correctly. 1`] = `
<v-container fluid="" data-cy="task-template-step-name">
  <h1 data-cy="wizard-title" class="font-weight-medium text--primary mb-8">
    <v-icon color="primary" size="48" class="mr-4">mdi-brush</v-icon>
    New Tasks
  </h1>
  <v_error_multiple error="[object Object]"></v_error_multiple>
  <p data-cy="wizard-name-subtitle" class="text--primary">
    The following steps will guide you on the creation of a new task group.
    Give a name to your Tasks:
  </p>
  <v-text-field label="Name" data-cy="task-template-name-input"></v-text-field>
  <wizard_navigation></wizard_navigation>
</v-container>
`;
